diff --git a/modules/hs_taxonomy.module b/modules/hs_taxonomy.module
index 8a591fc..d33ed87 100644
--- a/modules/hs_taxonomy.module
+++ b/modules/hs_taxonomy.module
@@ -55,7 +55,7 @@ function hs_taxonomy_form_field_ui_widget_type_form_alter(&$form, &$form_state)
       'root_term'                  => NULL,
       'entity_count_for_node_type' => NULL,
     );
-    $config_id = "taxonomy-$vid";
+    $config_id = "taxonomy-{$field['field_name']}";
     $defaults = array(
       // Enable the save_lineage setting by default if the multiple parents
       // vocabulary option is enabled.
@@ -109,7 +109,7 @@ function hs_taxonomy_form_field_ui_widget_type_form_alter(&$form, &$form_state)
 function hs_taxonomy_field_settings_submit(&$form, &$form_state) {
   $field = field_info_field($form['#field_name']);
   $vocabulary = taxonomy_vocabulary_machine_name_load($field['settings']['allowed_values'][0]['vocabulary']);
-  $config = hierarchical_select_common_config_get("taxonomy-$vocabulary->vid");
+  $config = hierarchical_select_common_config_get("taxonomy-{$field['field_name']}");
 
   if ($config['dropbox']['status'] || $config['save_lineage']) {
     $field = field_info_field($form['#field_name']);
@@ -131,8 +131,7 @@ function hs_taxonomy_form_field_ui_field_edit_form_alter(&$form, &$form_state) {
   if (isset($form['#field']['type']) && $form['#field']['type'] === 'taxonomy_term_reference' && $form['#instance']['widget']['type'] == 'taxonomy_hs') {
     require_once DRUPAL_ROOT . '/' . drupal_get_path('module', 'hierarchical_select') . '/includes/common.inc';
 
-    $vocabulary = taxonomy_vocabulary_machine_name_load($form['#field']['settings']['allowed_values'][0]['vocabulary']);
-    $config = hierarchical_select_common_config_get("taxonomy-$vocabulary->vid");
+    $config = hierarchical_select_common_config_get("taxonomy-{$form['#field']['field_name']}");
 
     if ($config['dropbox']['status'] || $config['save_lineage']) {
       $form['field']['cardinality']['#disabled'] = TRUE;
@@ -254,7 +253,7 @@ function hs_taxonomy_field_widget_form(&$form, &$form_state, $field, $instance,
     '#default_value' => $tids,
   );
 
-  hierarchical_select_common_config_apply($element, "taxonomy-$vocabulary->vid");
+  hierarchical_select_common_config_apply($element, "taxonomy-{$field['field_name']}");
 
   // Append another #process callback that transforms #return_value to the
   // format that Field API/Taxonomy Field expects.
@@ -338,7 +337,7 @@ function hs_taxonomy_field_formatter_prepare_view($entity_type, $entities, $fiel
 
   // Get the config for this field.
   module_load_include('inc', 'hierarchical_select', 'includes/common');
-  $config_id = "taxonomy-$vid";
+  $config_id = "taxonomy-{$field['field_name']}";
   $config = hierarchical_select_common_config_get($config_id);
   $config += array(
     'module' => 'hs_taxonomy',
@@ -690,7 +689,7 @@ function hs_taxonomy_hierarchical_select_config_info() {
             $bundles_info = field_info_bundles($entity_type);
             $entity_info = entity_get_info($entity_type);
             $vocabulary = taxonomy_vocabulary_machine_name_load($field['settings']['allowed_values'][0]['vocabulary']);
-            $config_id = "taxonomy-$vocabulary->vid";
+            $config_id = "taxonomy-{$instance['field_name']}";
             $config_info[$config_id] = array(
               'config_id'      => $config_id,
               'hierarchy type' => t('Taxonomy'),
