diff --git a/sites/all/modules/resource_conflict/resource_conflict.module b/sites/all/modules/resource_conflict/resource_conflict.module
index 6b885d7..c672903 100644
--- a/sites/all/modules/resource_conflict/resource_conflict.module
+++ b/sites/all/modules/resource_conflict/resource_conflict.module
@@ -144,7 +144,7 @@ function resource_conflict_form_alter(&$form, $form_state, $form_id) {
             $date_fields[$fieldname] = $field['label'];
           }
         }
-        elseif (isset($field['widget']['module']) && ($field['widget']['module'] == 'entityreference')) {
+        elseif (isset($field['widget']['module']) && ($field['widget']['module'] == 'entityreference') || ($field['display']['default']['module'] == 'taxonomy')) {
           $nodereference_fields[$fieldname] = $field['label'];
         }
       }
@@ -300,14 +300,22 @@ function _resource_conflict_get_node_resource_demand($node) {
        * file the error against the fake 'no_resources' element as CCK files
        * it's own "Illegal choice" error before we get called.
        */
-      if (is_numeric($reference['target_id'])) {
+      if (isset($reference['target_id'])) { 
+        $rid = $reference['target_id'];
+        $target_type = $fieldinfo['settings']['target_type'];
+      }
+      elseif (isset($reference['tid'])) {
+        $rid = $reference['tid'];
+        $target_type = 'taxonomy_term';
+      };
+      if (is_numeric($rid)) {
         $demand[] = array(
-          'target_id' => $reference['target_id'], 
-          'target_type' => $fieldinfo['settings']['target_type'],
-        );
+          'target_id' => $rid, 
+          'target_type' => $target_type,
         //$demand[$reference['target_id']] = TRUE;
+        );
       }
-      elseif (($reference != 'Add another item') && ($reference['target_id'] != '')) {
+      elseif (($reference != 'Add another item') && ($rid != '')) {
         form_set_error('no_resources', t('No bookable resources have been created. Please create resources to book before attempting to create a resource booking.'));
       }
     }
